### PR TITLE
Type container

### PR DESCRIPTION
**Pull request by eldadcohen1**
I created Type Container to resolve all the problems that we have with all the kinds of providers.
As : Image Source Provider and Data Source provider.
Today the container can only contain type with constructor without parameters.
all the type can b resolved just as a singleton right now. In the future there will be  a possibility for transient.

I made the relevant changes in the imagery package as well

### CODE REVIEW

File: src/app/app-models/map-source-providers/open-layers-TileWMS-source-provider.ts [hunk 1]
```diff
@@ -2,8 +2,7 @@
  * Created by AsafMas on 21/05/2017.
  */
 
-import * as ol from 'openlayers';
-import { BaseSourceProvider } from '@ansyn/map-source-provider';
+import { BaseSourceProvider } from '@ansyn/imagery';
```
Comment (asafMasa) [inline src/app/app-models/map-source-providers/open-layers-TileWMS-source-provider.ts]: import ol is missing

File: src/app/app-reducers/effects/map.app.effects.spec.ts [hunk 2]
```diff
@@ -12,6 +11,8 @@ import { Observable } from 'rxjs/Observable';
 import { ICasesState } from '@ansyn/menu-items/cases/reducers/cases.reducer';
 import { SelectCaseByIdAction } from '@ansyn/menu-items/cases/actions/cases.actions';
 import { configuration } from "configuration/configuration";
+import { TypeContainerService,TypeContainerModule} from '@ansyn/type-container';
```
Comment (asafMasa) [inline src/app/app-reducers/effects/map.app.effects.spec.ts]: outlining

File: src/app/app-reducers/effects/map.app.effects.spec.ts [hunk 3]
```diff
@@ -12,6 +11,8 @@ import { Observable } from 'rxjs/Observable';
...
+import { BaseSourceProvider } from '@ansyn/imagery/model/base-source-provider.model';
```
Comment (asafMasa) [inline src/app/app-reducers/effects/map.app.effects.spec.ts]: get from index

File: src/app/packages/type-container/index.ts [hunk 4]
```diff
@@ -0,0 +1,3 @@
+export { TypeContainerService, ITypeContainer } from './services/type-container.service';
+export { TypeContainerModule } from './type-container.module';
+export { RegisterOptions } from './models/register-options.model';
```
Comment (asafMasa) [inline src/app/packages/type-container/index.ts]: add new line at the end of the file

File: src/app/packages/type-container/models/type.model.ts [hunk 5]
```diff
@@ -0,0 +1,3 @@
+export interface Type extends Function {
+    new (...args: any[]): any;
+}
```
Comment (asafMasa) [inline src/app/packages/type-container/models/type.model.ts]: add new line at the end of the file

File: src/app/app-reducers/effects/map.app.effects.ts [hunk 6]
```diff
@@ -39,7 +40,7 @@ export class MapAppEffects {
 			const bboxPolygon = turf.bboxPolygon(bbox);
 			const extent = {topLeft: bboxPolygon.geometry.coordinates[0][0], topRight: bboxPolygon.geometry.coordinates[0][1], bottomLeft: bboxPolygon.geometry.coordinates[0][2], bottomRight:bboxPolygon.geometry.coordinates[0][3]};
 			const mapType = this.communicator.provideCommunicator(active_map_id).ActiveMap.mapType;
-			const sourceLoader = this.mapSourceProviderContainerService.resolve(mapType, overlay.sourceType);
+			const sourceLoader = this.typeContainerService.resolve(BaseSourceProvider,[mapType, overlay.sourceType].join(','));
 			sourceLoader.createAsync(overlay).then((layer)=> {
```
Comment (osadan) [inline src/app/app-reducers/effects/map.app.effects.ts]: please please not then

**Pull request merged**